Make Global and Profile settings inheritable

carlos-zamora opened the pull request:
## Summary of the Pull Request
Introduces `IInheritable` as an interface that helps move cascading settings into the Terminal Settings Model. `GlobalAppSettings` and `Profile` both are now `IInheritable`. `CascadiaSettings` was updated to `CreateChild()` for globals and each profile when we are loading the JSON data.

IInheritable does most of the heavy lifting. It introduces a two new macros and the interface. The macros help implement the fallback functionality for nullable and non-nullable settings.

## References
#7876 - Spec Addendum
#6904 - TSM Spec
#1564 - Settings UI

#7876 - `Copy()` needs to be updated to include _parent

carlos-zamora commented:
## TODO for this PR
- [ ] Update IDL with `Has...` functions
- [ ] Runtime error when `_globals->CreateChild()` is being assigned (only when loading settings.json, fine otherwise)
- [ ] Update spec and propagate changes from spec
- [ ] Profile has a few settings we handle manually (i.e. BI alignment). I need to hook those up properly too
- [ ] (Bonus Points) _maybe_ make these settings observable (we'll need that soon anyways)

DHowett commented on src/inc/til/coalesce.h:
```diff
@@ -22,6 +34,15 @@ namespace til
+    // Method Description:
+    // - Base case provided to throw an assertion if you call coalesce_value(opt, opt, opt)
+    template<typename T>
+    T coalesce_value(const winrt::Windows::Foundation::IReference<T>& base)
```
til cannot have an unguarded reference to `IReference`. `til` is conceptually below things that use winrt.

DHowett replied:
oh i see, you.. did the thing i did in JsonUtils. Clever.

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -88,14 +88,20 @@ Json::Value Profile::GenerateStub() const
+    if (_Name)
+    {
+        stub[JsonKey(NameKey)] = winrt::to_string(*_Name);
+    }
 
-    if (!_Source.empty())
+    if (_Source && !_Source.value().empty())
     {
-        stub[JsonKey(SourceKey)] = winrt::to_string(_Source);
+        stub[JsonKey(SourceKey)] = winrt::to_string(*_Source);
     }
 
-    stub[JsonKey(HiddenKey)] = _Hidden;
+    if (_Hidden)
+    {
+        stub[JsonKey(HiddenKey)] = *_Hidden;
+    }
```
stub is intended to _always contain these values_ -- make sure you always set them, even if they were inherited.

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,156 @@
+    struct IInheritable
+    {
+    public:
+        T* CreateChild() const
```
I'd rather you return a `com_ptr` from this; otherwise, there's a brief period where nobody owns the pointer and it can leak.

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,156 @@
+        T* CreateChild() const
+        {
+            auto child{ winrt::make_self<T>() };
+            child->_parent.attach(const_cast<T*>(static_cast<const T*>(this)));
```
does attach add a reference count properly?

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,156 @@
+        com_ptr<T> _parent{ nullptr };
+    };
+
+    // Nullable settings are settings that explicitly allow null to mean something
```
`// This is like std::optional, but we can use it in inheritance to determine whether the user explicitly cleared it`

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,156 @@
+    /* Dispatch event if value changed */                                   \
+    void Clear##name()                                                      \
+    {                                                                       \
+        if (Has##name())                                                    \
```
no need to check; cheap to assign nullopt over nullopt

carlos-zamora replied:
Ah. So the if-statement in "Clear" and the setter are for the _PropertyChanged dispatcher.

carlos-zamora replied:
Removed since we may not need these to be observable anyways. And making them observable shouldn't be too difficult to add later.

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,156 @@
+                   _##name :                                                \
+                   (_parent ?                                               \
+                        _parent->_get##name##Impl() :                       \
+                        type{ __VA_ARGS__ });                               \
```
no since you're doing the `__VA_ARGS__` thing in the real getter on line 61, change this to just return `nullopt`

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -590,6 +590,8 @@ winrt::com_ptr<CascadiaSettings> CascadiaSettings::FromJson(const Json::Value& j
 void CascadiaSettings::LayerJson(const Json::Value& json)
 {
+    // add a new inheritance layer, and apply json values to child
+    _globals.attach(_globals->CreateChild());
```
note that your inheritance here is incomplete! the baby globals has no color schemes, because they're not inherited... and if they were, you'd want it to be a .. copy? it's hard to have a layered array, for sure.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -627,10 +629,14 @@ void CascadiaSettings::LayerJson(const Json::Value& json)
+        auto parentProj{ _profiles.GetAt(*profileIndex) };
+        auto parent{ winrt::get_self<Profile>(parentProj) };
+        auto childImpl{ parent->CreateChild() };
+        childImpl->LayerJson(profileJson);
```
don't you need to.. like, overwrite the parent in the array with the child now that it's inherited?

carlos-zamora replied:
haha. I did all the work of getting the index, then just didn't use it

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -151,22 +157,22 @@ bool Profile::ShouldBeLayered(const Json::Value& json) const
     {
         if (otherHadSource)
         {
             // If we have a source and the other has a source, compare them!
-            sourceMatches = *otherSource == _Source;
+            sourceMatches = *otherSource == *_Source;
         }
         else
         {
             // Special case the legacy dynamic profiles here. In this case,
             // `this` is a dynamic profile with a source, and our _source is one
             // of the legacy DPG namespaces. We're looking to see if the other
             // json object has the same guid, but _no_ "source"
-            if (_Source == WslGeneratorNamespace ||
-                _Source == AzureGeneratorNamespace ||
-                _Source == PowershellCoreGeneratorNamespace)
+            if (*_Source == WslGeneratorNamespace ||
+                *_Source == AzureGeneratorNamespace ||
+                *_Source == PowershellCoreGeneratorNamespace)
```
careful with source. source may need to be a "see-through" property. Gotta handle it appropriately.

carlos-zamora commented:
This branch still doesn't work right, and I can't figure out why. It builds/deploys, but...
- "Windows PowerShell" is still a profile, but it shouldn't be for me. It also has the name in "profile.defaults"
- "Command Prompt" doesn't have an icon
- "Raspberry Pi" custom profile seems to be fine
- "Ubuntu" has name from "profile.defaults"
- "PowerShell" isn't recognized as the default profile (and has name from "profile.defaults")

![dropdown menu](https://user-images.githubusercontent.com/11050425/96536244-797b4880-1248-11eb-8e6e-b3808bbb056b.png)

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,215 @@
+
+        void InsertParent(size_t index, com_ptr<T> parent)
+        {
+            auto pos{ _parents.begin() };
+            for (size_t i=0; i<index; ++i)
+            {
+                ++pos;
+            }
```
you can probably do `_parents.begin() + index`

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -0,0 +1,215 @@
+    template<typename T>
+    struct NullableSetting
+    {
+        winrt::Windows::Foundation::IReference<T> setting{ nullptr };
```
if you use `std::optional` even inside NullableSetting -- for the same exact reason as you'd use it outside -- you don't need a special string type. Remember: IReference can be our _api contract_ but it doesn't need to pervade our internal implementations

carlos-zamora replied:
Problem is, then we'd have to convert between `nullopt` and `nullptr`. I'd rather just keep it saved as an `IReference` internally. That way, when we get/set the value (and it's valid) we just return `NullableSetting.setting`, instead of translating it on the way in/out.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -639,13 +654,13 @@ void CascadiaSettings::_LayerOrCreateProfile(const Json::Value& profileJson)
         {
-            auto profile = winrt::make_self<Profile>();
+            auto profile{ winrt::make_self<Profile>() };
 
             // GH#2325: If we have a set of default profile settings, apply them here.
             // We _won't_ have these settings yet for defaults, dynamic profiles.
             if (_userDefaultProfileSettings)
             {
-                profile->LayerJson(_userDefaultProfileSettings);
+                profile = _userDefaultProfileSettings->CreateChild();
```
careful - this will create a useless profile and then delete it every time

you probably want `InsertParent(0`

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.h:
```diff
@@ -130,7 +127,6 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
     private:
-        std::optional<winrt::guid> _Guid{ std::nullopt };
         std::optional<winrt::guid> _ConnectionType{ std::nullopt };
```
make this one heritable as well! otherwise ACS won't work for the same reason GUID didn't work.

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -397,6 +427,52 @@ bool Profile::IsDynamicProfileObject(const Json::Value& json)
+// - profile: the Profile to be updated
+// Return Value:
+// - <None>
+void Profile::ApplyTo(Profile* profile) const
```
reminder to scrub applyto! :smile:

carlos-zamora replied:
Idea: using `ApplyTo()` to make `Copy()` easier. Less repeated code!

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -347,18 +345,12 @@ winrt::hstring Profile::ExpandedBackgroundImagePath() const
-    const auto path{ StartingDirectory() };
-    const auto pathString{ path.as<winrt::Windows::Foundation::IReference<winrt::hstring>>() };
-    if (pathString == nullptr)
+    if (_StartingDirectory)
```
don't we need to check the inheritance tree, in case the user set it in `p.defaults`?

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -632,14 +632,13 @@ void CascadiaSettings::_LayerOrCreateProfile(const Json::Value& profileJson)
+        // We don't actually need to CreateChild() here.
+        // When we loaded Profile.Defaults, we created an empty child already.
+        // So this just populates the empty child
+        parent->LayerJson(profileJson);
```
this is only guaranteed _if we had a `p.d.`!!!_ otherwise it will layer into the existing profiles!

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp:
```diff
@@ -86,6 +81,26 @@ winrt::Microsoft::Terminal::Settings::Model::CascadiaSettings CascadiaSettings::
+
+    // Our profiles inheritance graph doesn't have a formal root.
+    // However, if we create a dummy Profile, and set _profiles as the parent,
+    //  we now have a root. So we'll do just that, and
```
AND WHAT! I'm gripped!

DHowett replied:
(also i expressed this to Carlos in IM, but this is _very clever_)

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -70,60 +70,106 @@ Profile::Profile(guid guid) :
+// - visited - a map of which Profiles have been visited, and, if so, a reference to the Profile's clone
+// Return Value:
+// - a clone in both inheritance structure and Profile values of sourceGraph
+winrt::com_ptr<Profile> Profile::CloneInheritanceGraph(winrt::com_ptr<Profile> sourceGraph, winrt::com_ptr<Profile> cloneGraph, std::unordered_map<void*, winrt::com_ptr<Profile>> visited)
```
this is expensive and incorrect! you're copying the map at every step, so every time you recurse you're adding nodes to a map that gets deleted.

This has to come in as a reference (`map&`) so every node at every level can see the whole visited list.

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -70,60 +70,106 @@ Profile::Profile(guid guid) :
+        {
+            // If we visited this Profile already...
+            auto kv{ visited.find(sourceParent.get()) };
+            if (visited.find(sourceParent.get()) != visited.end())
```
presumably, 
```
if(kv != visited.end())
```

carlos-zamora commented:
Other than changing IReference to std::optional. This PR is good to go now.

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -438,42 +507,45 @@ winrt::guid Profile::GetGuidOrGenerateForJson(const Json::Value& json) noexcept
+    }
 }
 
-winrt::guid Profile::Guid() const
+const VerticalAlignment Profile::BackgroundImageVerticalAlignment() const noexcept
 {
-    // This can throw if we never had our guid set to a legitimate value.
-    THROW_HR_IF_MSG(E_FAIL, !_Guid.has_value(), "Profile._guid always expected to have a value");
-    return *_Guid;
+    const auto val{ _getBackgroundImageAlignmentImpl() };
+    return val ? std::get<VerticalAlignment>(*val) : VerticalAlignment::Center;
 }
 
-void Profile::Guid(const winrt::guid& guid) noexcept
+void Profile::BackgroundImageVerticalAlignment(const VerticalAlignment& value) noexcept
 {
-    _Guid = guid;
+    if (!HasBackgroundImageAlignment() || std::get<VerticalAlignment>(*_BackgroundImageAlignment) != value)
+    {
+        std::get<VerticalAlignment>(*_BackgroundImageAlignment) = value;
+    }
 }
+#pragma endregion
```
you can't do this.

the optional's contents are _UNDEFINED_ if it doesn't have a value. Remember last time? :smile:

DHowett commented on src/inc/til/coalesce.h:
```diff
@@ -3,8 +3,20 @@
 
 #pragma once
 
+namespace winrt
+{
+    // If we don't use winrt, nobody will include the ConversionTraits for winrt stuff.
+    // If nobody includes it, these forward declarations will suffice.
+    namespace Windows::Foundation
+    {
+        template<typename T>
```
i think you can revert til if you remove the ireference stuff

DHowett reviewed (changes requested):
-

zadjii-msft commented on src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp:
```diff
@@ -85,10 +80,54 @@ winrt::Microsoft::Terminal::Settings::Model::CascadiaSettings CascadiaSettings::
+// - cloneSettings: the CascadiaSettings we're copying the inheritance tree to
+// Return Value:
+// - <none>
+void CascadiaSettings::_CopyProfileInheritanceTree(CascadiaSettings* cloneSettings) const
```
Is there a reason this is a `CascadiaSettings*`? That seems dangerous to me. Shouldn't that be like, a `com_ptr<>&` or something?

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Profile.h:
```diff
@@ -57,83 +61,101 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         void GenerateGuidIfNecessary() noexcept;
         static guid GetGuidOrGenerateForJson(const Json::Value& json) noexcept;
 
-        bool HasGuid() const noexcept;
```
Wait so we can get rid of the `HasGuid` and `optional<GUID>` now because `GETSET_SETTING` is already basically doing this for us?

carlos-zamora replied:
Correct. `GETSET_SETTING` introduces...
- `HasGuid()`
- `Guid()` (getter)
- `Guid(guid)` (setter)
- `ClearGuid()`
as well as the optional member and a helper function that does most of the work for the getter

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -169,7 +227,9 @@ winrt::com_ptr<winrt::Microsoft::Terminal::Settings::Model::implementation::Prof
-    if (!_Guid.has_value())
+    // guid was not set
+    const auto myGuid{ Guid() };
+    if (myGuid == winrt::guid{})
```
I'm pretty sure there's a difference between not having a GUID and having the default-constructed GUID as a value. IIRC, the default-constructed one is all zeros, and that _is_ a valid GUID setting!

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -247,10 +308,10 @@ void Profile::LayerJson(const Json::Value& json)
-    JsonUtils::GetValueForKey(json, BackgroundKey, _Background);
-    JsonUtils::GetValueForKey(json, SelectionBackgroundKey, _SelectionBackground);
-    JsonUtils::GetValueForKey(json, CursorColorKey, _CursorColor);
+    _Foreground.set = JsonUtils::GetValueForKey(json, ForegroundKey, _Foreground.setting);
```
Do we want to file a follow-up for enlightening `JsonUtils` to be able to handle this case just as gracefully as the others?

carlos-zamora replied:
CC @DHowett 
I'm not sure what to do here. I've hit a few times where JsonUtils needed a little tweak. At the same time, they've been pretty small things so idk if complicating JsonUtils more would be the right approach. But at the same time, eh, it couldn't hurt?

DHowett replied:
That was this discussion (transcribed from Teams markup...)

1. remove optional deduction from `Get/SetValue/ForKey`'s `Converter` detector
2. implement a partially specialized `ConversionTrait` over `std::optional<T>`
   1. `template<typename T> struct ConversionTrait<std::optional<T>>` may not even be legal code
   2. implement `FromJson` as return `nullopt`; or whatever if it's `null`
3. move the null detection logic from `GetValue` into `ConversionTrait<std::optional<T>>`
4. delete all of the `DeduceOptional` code
5. implement a `ConversionTrait<IReference<T>>` that is identical to the `std::optional<T>` version
6. mentally accept the fact that putting `null` in a non-nullable field will now cause a deserialization type error ("expected string, got null")
   1. mentally accept the fact that `std::optional<int>` will say, like, "expected null or number" which is cool
7. ------ NOW YOU CAN START TO FIX YOUR ORIGINAL PROBLEM -------
8. introduce `Settings::ExplicitlyNullable<T>` as `struct { bool _set; std::optional<T> _actualValue; }`
9. Implement `ConversionTrait<ExplicitlyNullable<T>>`
10. in `FromJson`, set `_set` even if it's null
11. then you can just check `_set`

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -70,49 +70,106 @@ Profile::Profile(guid guid) :
+    if (visited.find(sourceGraph.get()) == visited.end() && !sourceGraph->_parents.empty())
+    {
+        // iterate through all of our parents to copy them
+        for (auto sourceParent : sourceGraph->_parents)
```
```suggestion
        for (const auto& sourceParent : sourceGraph->_parents)
```
maybe?

zadjii-msft commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -511,7 +511,7 @@ bool CascadiaSettings::_AppendDynamicProfilesToUserSettings()
     for (const auto& profile : _profiles)
     {
-        if (!profile.HasGuid())
+        if (profile.Guid() == winrt::guid{})
```
Same thing here - `guid{}` != `nullopt`

zadjii-msft reviewed (changes requested):
Hey I'm at 12/17 but it's also nearly the EOD here and I can see there are some new commits, so this is what I've got so far.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -511,20 +511,6 @@ bool CascadiaSettings::_AppendDynamicProfilesToUserSettings()
-            // would be impossible to be layered.
-            continue;
-        }
-
```
this might change the behavior of guid-less dynamic profiles. is that acceptable? we don't think any exist today.

carlos-zamora replied:
I think it is? We get a collision when two profiles have...
- the same guid
- the same name & source --> same deterministically auto-generated guid

This seems relevant to proto-extensions. Authors need to claim a guid. Or (on our end) we need to dig deeper into what the "source" is for a proto-extension (not just "local"). Both seem reasonable to me.

DHowett commented on src/cascadia/TerminalSettingsModel/IInheritable.h:
```diff
@@ -49,7 +52,13 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
-        std::vector<com_ptr<T>> ExportParents()
+        // Method Description:
+        // - Exports list of parents as an immutable view.
```
not a view, not immutable. contractually "not going to change my copy", but you can still wreak havoc with the parents inside the list

carlos-zamora commented:
@zadjii-msft So, I moved guid generation to the getter like this:
```c++
        GETSET_SETTING(guid, Guid, _GenerateGuidForProfile(Name(), Source()));
```
So now, when you have a `Profile` and you ask for its Guid, we check the inheritance tree for a Guid that's defined. If none were defined, we _deterministically_ generate one for this `Profile` based on the name & source (same as before). Basically, I moved this logic into the getter, as opposed to generating one when we are validating.

It works. It even fixed a bug in a test. I think this is the behavior we want when we expose inheritance in the JSON too?

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -227,40 +221,34 @@ winrt::com_ptr<winrt::Microsoft::Terminal::Settings::Model::implementation::Prof
     const auto mySource{ Source() };
     if (!mySource.empty())
     {
-        if (otherHadSource)
```
there's a very subtle difference between what you have now and what you had before regarding the treatment of `null` in the source field in json.

also, i think it would have resulted in slightly undefined behavior if you had previously had a `null` source in json.

carlos-zamora commented:
I still have 2 failing tests: Unbind/Rebind nested commands. I'll probably look into them this afternoon.

I want to get a head start on serialization though. So I'm taking that on today, and that'll be based off this branch. I have a feeling I may need @zadjii-msft 's help with the two failing tests anyways.

zadjii-msft reviewed (changes requested):
Alright, I have a few more questions before I'm ready to sign off, but I think this is close

carlos-zamora commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -70,49 +70,103 @@ Profile::Profile(guid guid) :
+    profile->_CursorHeight = source->_CursorHeight;
+    profile->_BellStyle = source->_BellStyle;
+
+    // copy settings that did not use the GETSET macro
```
I did this to help me make sure I copied everything over. I can remove it, if you want, since it doesn't add anything.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -55,14 +55,42 @@ static constexpr bool debugFeaturesDefault{ false };
+    {
+        _keymap = parent->_keymap->Copy();
+        std::copy(parent->_keybindingsWarnings.begin(), parent->_keybindingsWarnings.end(), std::back_inserter(_keybindingsWarnings));
+        for (auto kv : parent->_colorSchemes)
```
Tried 😟. We need a newer version of CppWinRT for that to work (I cri every tim)

zadjii-msft replied:
Oh well _I'm on it_

carlos-zamora commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -55,14 +55,42 @@ static constexpr bool debugFeaturesDefault{ false };
+// - <none>
+// Return Value:
+// - <none>
+void GlobalAppSettings::_FinalizeInheritance()
```
> So, to get this right, every node in the GlobalAppSettings inheritance graph has a full copy of the parent's schemes and commands, correct?

Minor thing I messed up. These should be `std::move` instead of copies. That way we don't hold on to an extra copy of color schemes, commands, etc. that we'll never touch. So instead we'll just have one copy of all of this and pass it down to the leaf.

>  So for these, we can't really tell what layer a command was actually defined at, right? But there's no other way of doing this, because we can't easily express "Layer 2 undefined {command} from layer 1", right?

Exactly. This goes back to the discussion we were having about doing action IDs sooner than later. With action IDs, we should be able to layer an action onto our ancestors. Additionally, we'll have to design our own data structure solution that...
- checks the leaf for an action
- if none are found, checks the parent's
while also maintaining more complex functionality like "unbind" and "rebind" a command. These two work items are pretty intertwined IMO.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -55,14 +55,42 @@ static constexpr bool debugFeaturesDefault{ false };
+{
+    // Globals only ever has 1 parent
+    FAIL_FAST_IF(_parents.size() > 1);
+    for (auto parent : _parents)
```
Actually, the loop looks much cleaner. Because you could either have 0 or 1 parent. Leaving this one as is.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -627,10 +615,28 @@ void CascadiaSettings::LayerJson(const Json::Value& json)
+        auto parentProj{ _profiles.GetAt(*profileIndex) };
+        auto parent{ winrt::get_self<Profile>(parentProj) };
+
+        if (_userDefaultProfileSettings)
```
The Profile Inheritance Graph should look like this (layer settingsA onto settingsB):
```
[A]          [B]          [prof.def]
bar.exe ---> foo.exe ---> cmd.exe 
                     \--> null Profile
```

But because we hold one reference to `_userDefaultProfileSettings`, you're right, we end up doing this:
```
[A]          [prof.def]   
bar.exe ---> cmd.exe
        \--> foo.exe [B] ---> cmd.exe [prof.def] 
                         \--> null Profile
```

So I think we may need to make `CascadiaSettings` inheritable to get around that issue? But yeah, this'll only really be a problem when we allow multiple files being layered (settings.json and defaults.json don't count because there's special handling for defaults.json; we know there isn't a profiles.defaults in defaults.json).

carlos-zamora commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -55,14 +55,42 @@ static constexpr bool debugFeaturesDefault{ false };
+void GlobalAppSettings::_FinalizeInheritance()
+{
+    // Globals only ever has 1 parent
+    FAIL_FAST_IF(_parents.size() > 1);
```
From Dustin:
> having multiple parents is a coding error that we must fix; a failfast is acceptable
> if this hits in the wild, our understanding of our settings model is fundamentally incorrect

carlos-zamora commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -153,7 +236,10 @@ winrt::com_ptr<GlobalAppSettings> GlobalAppSettings::FromJson(const Json::Value&
+    // _validDefaultProfile keeps track of whether we've verified that DefaultProfile points to something
+    // CascadiaSettings::_ResolveDefaultProfile performs a validation and updates DefaultProfile() with the
+    // resolved value, then making it valid.
+    _validDefaultProfile = !JsonUtils::GetValueForKey(json, DefaultProfileKey, _UnparsedDefaultProfile);
```
Right. Changed this to:
```c++
    _validDefaultProfile = false;
    JsonUtils::GetValueForKey(json, DefaultProfileKey, _UnparsedDefaultProfile);
```
`_validDefaultProfile` _needs_ to be reset to false when we import a new value via the json. This is because `_defaultProfile` is not populated yet (only `_UnparsedDefaultProfile`). In `CascadiaSettings`, we resolve/validate the provided value in `ResolveDefaultProfile` and `ValidateDefaultProfileExists`. So at this point, `_defaultProfile` can be extracted and used properly.

For the purposes of the Settings UI, we should only ever interact with the `UnparsedDefaultProfile` to give the user the flexibility of using a guid or name as an identifier. Eventually, we'll make it so that SUI just gives you a list of profiles to choose from, and this extra level of complication can maybe go away with a refactor?

DHowett commented:
must-fix:

* [ ] default profile fails to resolve on empty settings case

pleasant to have:

* [ ] if the p.d block is empty, don't even make a node for it!

DHowett commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -113,19 +113,28 @@ winrt::com_ptr<GlobalAppSettings> GlobalAppSettings::Copy() const
+        for (auto kv : _colorSchemes)
+        {
+            const auto schemeImpl{ winrt::get_self<ColorScheme>(kv.Value()) };
+            globals->_colorSchemes.Insert(kv.Key(), *schemeImpl->Copy());
+        }
     }
 
-    for (auto kv : _commands)
+    if (_commands)
     {
-        const auto commandImpl{ winrt::get_self<Command>(kv.Value()) };
-        globals->_commands.Insert(kv.Key(), *commandImpl->Copy());
+        for (auto kv : _commands)
+        {
+            const auto commandImpl{ winrt::get_self<Command>(kv.Value()) };
+            globals->_commands.Insert(kv.Key(), *commandImpl->Copy());
+        }
```
what happened to std::move...

DHowett commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.h:
```diff
@@ -52,43 +54,49 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         Windows::Foundation::Collections::IMap<hstring, Model::ColorScheme> _colorSchemes;
         Windows::Foundation::Collections::IMap<hstring, Model::Command> _commands;
 
+        std::optional<hstring> _getUnparsedDefaultProfileImpl() const;
```
how is this different from `GETSET_SETTING`?

carlos-zamora replied:
The setter for UnparsedDefaultProfile needs to set `_validDefaultProfile` to false. We need this to validate that the provided "defaultProfile" value actually maps to something.

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -169,50 +221,47 @@ winrt::com_ptr<winrt::Microsoft::Terminal::Settings::Model::implementation::Prof
             // Special case the legacy dynamic profiles here. In this case,
             // `this` is a dynamic profile with a source, and our _source is one
             // of the legacy DPG namespaces. We're looking to see if the other
             // json object has the same guid, but _no_ "source"
-            if (_Source == WslGeneratorNamespace ||
-                _Source == AzureGeneratorNamespace ||
-                _Source == PowershellCoreGeneratorNamespace)
+            if (mySource == WslGeneratorNamespace ||
+                mySource == AzureGeneratorNamespace ||
+                mySource == PowershellCoreGeneratorNamespace)
             {
```
FYI @carlos-zamora  this is the code that makes "source" optional for the pre-0.5 dynamic generators

DHowett reviewed (approved):
Let's do this.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/Profile.h:
```diff
@@ -73,10 +69,11 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         void BackgroundImageVerticalAlignment(const Windows::UI::Xaml::VerticalAlignment& value) noexcept;
 
-        GETSET_SETTING(guid, Guid);
+        GETSET_SETTING(guid, Guid, _GenerateGuidForProfile(Name(), Source()));
```
I'm _shocked_ that this works, but I get why it does

zadjii-msft reviewed (approved):
Good work landing all this, it's a complicated change in a complicated codebase, but looks like it should work. I'm sure we'll find out very fast if it doesn't 😄
